### DISCUSSION TITLE
RadioLib & LowPowerLib & 328P not liking each other

### DISCUSSION THREAD

**Discussion by Rustie0125** [jgromes/RadioLib#732] [General]
Hey All. 

Wanted to see who could shed some wisdom here. 

I have RadioLib working with Sx1262 and example code. Check
I have LowPowerLib working with Periodic wakeup and flashing LED afterwakeup. Check

Combine the two and its Silence in the night, The Serial printouts are garbage after wakeup and LED no longer flashes. for context Digital pin 5 is a load switch that turns the Radio power off after transmission and back on before next round. I suspect the issues is some systems that do not get turned back on after sleep period. but using the SPI_ON, UART0_ON commands seems to have no effect, Alos tried moving the initialization into the main loop to setup the system after sleep again but no action.  So I wanted to see if anyone can spot anything obvious or recommend another lowpower Library I can use with RadioLIB

```c++
#include "LowPower.h"
#include <RadioLib.h>
unsigned long Wake_Counter =0;

SX1262 radio = new Module(10, 2, 3, 9);

void setup()
{
  Serial.begin(9600);
  radio.begin(917.5, 250.0, 12, 5, 0x18, 10, 8);
  pinMode(LED_BUILTIN,OUTPUT);
  pinMode(5,OUTPUT);
  digitalWrite(LED_BUILTIN,LOW);    
}

void loop() 
{
    // Enter power down state for 8 s with ADC and BOD module disabled
    LowPower.powerDown(SLEEP_8S, ADC_OFF, BOD_OFF); 
    Serial.begin(9600);
    //Wake_Counter++;
    //if ( Wake_Counter > 2)
    //{
    digitalWrite(5,HIGH);
    radio.begin(917.5, 250.0, 12, 5, 0x18, 10, 8);
    radio.transmit("Hello World!");
    digitalWrite(LED_BUILTIN,HIGH);
    delay(1000);
    digitalWrite(LED_BUILTIN,LOW);
    digitalWrite(5,LOW);
    //Wake_Counter = 0;
    //}
    Serial.println(Wake_Counter);
}
```

**Comment by jgromes** [comment]
It's been quite some time since I last used LowPowerLib with RadioLib, so it is possible something is broken. We used this combination on FOSSASAT-1 without problems though.

However, the fact that Serial output is also garbled seems to point to some issue outside of RadioLib - maybe try without RadioLib, just Serial print after waking up?

**Reply by Rustie0125** [reply]
Hey, 

Yes I managed to fix tye issues with Serial not printing by adding small time delays but I'm still having issues with sleep current. 

When the device boots its sees 4uA sleep current. 2 mon later after the first succesfull TX. The sleep current is stuck at 500uA. If I remove RadioLib then it's 4uA again. I suspect SPI communication is not put to sleep after tx or blocking the low power library from doing what it needs to. 

If I use lowpower Idle state and manually try turn off SPI. Then my sleep current is at 1.2mA. 

So seems like Radiolib and low power SPI settings not playing nicely.

**Reply by jgromes** [reply]
> for context Digital pin 5 is a load switch

How exactly is it connected? From the code you have above, it seems to me you have to pull 5 high to enable power to the radio. But then again, you are not changing the state of that pin prior to the first `radio.begin()` call, which seems strange.

>  I suspect SPI communication is not put to sleep after tx or blocking the low power library from doing what it needs to

Well, it's not. RadioLib doesn't change the state of the SPI peripheral at all, and it's not supposed to. Internally, each time you call `radio.begin()`, it calls `SPI.begin()`, since it's assumed the begin method is only called at startup. If you have to disasble the SPI bus after using it, you will have to do it manually (e.g. by calling `SPI.end()` before going to sleep).

**Comment by Rustie0125** [comment]
Hi Yes the code linked above was a couple of version ago. here is the latest attempt, I know Radio lib was not ment to stop the SPI comms, I was suggesting Radiolib stops lowpower lib from shutting down the SPI bus somehow. 

Ill try the SPI.end() and see if that makes a difference. 
Thank you.
```
`#include <LowPower.h>


#include <RadioLib.h>

// SX1262 has the following connections:
// NSS pin:   10
// DIO1 pin:  2
// NRST pin:  3
// BUSY pin:  9

int Counter =0;
int runcount=0;
int VbatOK;
// or using RadioShield
// https://github.com/jgromes/RadioShield
//SX1262 radio = RadioShield.ModuleA;

// or using CubeCell
//SX1262 radio = new Module(RADIOLIB_BUILTIN_MODULE);
float Sensor1Data;
int   Sensor2Data;
int txdata;

void setup() {
 Serial.begin(9600);
 pinMode(5,OUTPUT);
 pinMode(6,INPUT);
 pinMode(A0,INPUT);
 digitalWrite(5,LOW);
 analogReference(INTERNAL);
  // initialize SX1262 with default settings
  //Serial.print(F("[SX1262] Initializing ... "));
  //radio.begin(907.5, 500.0, 7, 5, 0x31, 10, 8);
  digitalWrite(5,LOW);
  digitalWrite(10,LOW);
  digitalWrite(11,LOW);
  digitalWrite(12,LOW);
  digitalWrite(13,LOW);
}

void loop() 
{  
  LowPower.powerDown(SLEEP_8S, ADC_OFF, BOD_OFF);
  //delay(10);
   if(Counter >= 20)
   {
     delay(100);
     pinMode(6,INPUT);
     VbatOK = digitalRead(6);
     if( VbatOK == HIGH)
     {
     digitalWrite(5,HIGH);
     delay(100);
     SX1262 radio = new Module(10, 2, 3, 9);
     delay(100);
     radio.begin(907.5, 500.0, 7, 5, 0x31, 10, 8);
     delay(300);
     float SuperCap = ((analogRead(A0))*0.004321);
     String myData = String(SuperCap)+ " mV " + String(runcount)+ " Cycles";
     radio.transmit(myData);
     //digitalWrite(5,LOW);
     }
     Counter = 0;
     runcount++;
     //pinMode(6,OUTPUT);
     //digitalWrite(6,LOW);
     digitalWrite(5,LOW);
     digitalWrite(10,LOW);
     digitalWrite(11,LOW);
     digitalWrite(12,LOW);
     digitalWrite(13,LOW);
   }
   Counter++;
}`
```

**Reply by Rustie0125** [reply]
No luck calling SPI.end before going to sleep again. Stuck on 512uA  
after first TX. System works as ex0ected just sleep current high. When booting the 4uA us seen so not a static load somewhere on a pin.

**Comment by Rustie0125** [comment]
Ya I don't get it. this is the latest version, 500uA sleep current and also skips on transmissions Pin 6 is tied to vcc so always high. delays added everywhere.  One thing I forgot to mention the device is running custom bootloader from minicore 
<img width="163" alt="image" src="https://user-images.githubusercontent.com/27943282/234124743-cd618af9-72ee-46b1-9229-75bb135111f8.png">


```c++
#include <LowPower.h>
#include <RadioLib.h>

unsigned long Counter =0;
unsigned long runcount=0;
int VbatOK;

void setup() 
{
 //Serial.begin(9600);
  pinMode(5,OUTPUT);
  pinMode(6,INPUT);
  pinMode(A0,INPUT);
  digitalWrite(5,LOW);
  analogReference(INTERNAL);
  digitalWrite(5,LOW);
  digitalWrite(10,LOW);
  digitalWrite(11,LOW);
  digitalWrite(12,LOW);
  digitalWrite(13,LOW);
}

void loop() 
{  
  
  delay(100);
   if(Counter > 20)
   {
     //delay(100);
     pinMode(5,OUTPUT);
     pinMode(6,INPUT);
     pinMode(A0,INPUT);
     delay(100);
     VbatOK = digitalRead(6);
     if( VbatOK == HIGH)
     {
     digitalWrite(5,HIGH);
     delay(200);
     SX1262 radio = new Module(10, 2, 3, 9);
     delay(200);
     radio.begin(907.5, 500.0, 7, 5, 0x31, 10, 8);
     delay(100);
     float SuperCap = ((analogRead(A0))*4.39); // Resistor Ratio 680K/220K
     String myData = String(SuperCap)+ " mV " + String(runcount)+ " Cycles";
     radio.transmit(myData);
     delay(350);
     digitalWrite(5,LOW);
     SPI.end();
     delay(100);
     }
     Counter = 0;
     runcount++;
   }
   Counter++;
   LowPower.powerDown(SLEEP_8S, ADC_OFF, BOD_OFF);
}
```

**Reply by jgromes** [reply]
This was kind of bugging me, so I just installed LowPowerLib, and in the SX126x transmit example I put `LowPower.powerDown(SLEEP_8S, ADC_OFF, BOD_OFF);` instead of the original `delay(1000);` - and everything seems to be working, I see regular LoRa transmissions after 8 seconds. I didn't measure the power draw, but the functionality seems to be correct.

Also - no delays needed to get Serial working. I'm using a plain Arduino Uno, so maybe the bootloader makes a difference?

**Comment by Rustie0125** [comment]
Mmm good point.

Thanx for testing this. I would expect the functions just to take longer at
slower clock rate but delays don't seem to have an effect.

I'm running the pro mini at 1.8V BOD and 1mhz becuase this is for energy
harvesting project so trying to reduce the runtime current. So I can test
to confirm with normal bootloader but that will really limit the effective
operating range of the storage element.

**Reply by Rustie0125** [reply]
No Luck.. 

Loaded the Standard Arduino 8mhz 3v3 bootloader back and tried unmodified pro mini, Same result. bizarre I have done this before and it was fine.  What I don't is the device goes into the immediate 4uA sleep mode as expected as the if condition is not true. but after the first 3 min when it becomes true the device is stuck on 500uA between transmissions.

**Comment by Rustie0125** [comment]
Iv been looking for advice over on Arduino forums and someone seems to have objection in leaving the SPI lines physicals connected while turning the radio power on and off. if they both run at the same voltage and both use the same ground for reference I don't see the issue electrically ? it just means the peripherals need to be setup every time you wake up.. Does the Radio Lib library allow for low power state commands to the Semtech chips
